dozer-api/src/cache_builder/mod.rs

```diff
@@ -1,295 +1,299 @@
-    describe_histogram!(
-        DATA_LATENCY_HISTOGRAM_NAME,
-        "End-to-end data latency in seconds"
-    );
+        // Create log reader.
+        let log_reader_options = get_log_reader_options(endpoint);
+        let log_reader = create_log_reader(
+            &mut client,
+            endpoint_meta.schema,
+            log_reader_options.clone(),
+            state.next_log_position(),
+        )
+        .await?;
 
-    const OPERATION_TYPE_LABEL: &str = "operation_type";
-    const SNAPSHOTTING_LABEL: &str = "snapshotting";
+        Ok((
+            Self {
+                client,
+                cache_manager,
+                labels: labels.labels().clone(),
+                cache_write_options,
+                serving,
+                state,
+                log_reader_options,
+                log_reader,
+            },
+            endpoint_meta.descriptor_bytes,
+        ))
+    }
 
-    let mut snapshotting = !cache.is_snapshotting_done()?;
+    pub fn cache_reader(&self) -> &Arc<ArcSwap<CacheReader>> {
+        &self.serving
+    }
 
-    while let Some(op_and_pos) = receiver.blocking_recv() {
-        progress_bar.set_position(op_and_pos.pos + 1);
-        match op_and_pos.op {
-            LogOperation::Op { op } => match op {
-                Operation::Delete { old } => {
-                    if let Some(meta) = cache.delete(&old)? {
-                        if let Some((endpoint_name, operations_sender)) = operations_sender.as_ref()
-                        {
-                            let operation = types_helper::map_delete_operation(
-                                endpoint_name.clone(),
-                                CacheRecord::new(meta.id, meta.version, old),
-                            );
-                            send_and_log_error(operations_sender, operation);
-                        }
-                    }
-                    let mut labels = cache.labels().clone();
-                    labels.push(OPERATION_TYPE_LABEL, "delete");
-                    labels.push(SNAPSHOTTING_LABEL, snapshotting_str(snapshotting));
-                    increment_counter!(CACHE_OPERATION_COUNTER_NAME, labels);
-                }
-                Operation::Insert { new } => {
-                    let result = cache.insert(&new)?;
-                    let mut labels = cache.labels().clone();
-                    labels.push(OPERATION_TYPE_LABEL, "insert");
-                    labels.push(SNAPSHOTTING_LABEL, snapshotting_str(snapshotting));
-                    increment_counter!(CACHE_OPERATION_COUNTER_NAME, labels);
+    pub fn run(
+        mut self,
+        runtime: Arc<Runtime>,
+        mut cancel: impl Future<Output = ()> + Unpin + Send + 'static,
+        operations_sender: Option<(String, Sender<GrpcOperation>)>,
+    ) -> Result<(), CacheError> {
+        loop {
+            let Some(read_one_result) = runtime.block_on(self.read_one_with_cancel(cancel)) else {
```

Comment:
`runtime.block_on()` on the hot path will be slower than the previous solution that ran the log reading in a worker (https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#non-worker-future), which was a bit cleaner as well.

I think I'd like to invert control here, and have an async task which orchestrates cache building workers, swapping these tasks out when the log gets swapped. Cancellation would also be handled in the orchestrator task, as would swapping the `serving` cache.

The worker would get its ops through a channel, which has the added benefit of making the worker purely sync and not require a `runtime`.

Reply:
There were two reasons to implement it like this:

1. The `LogReader` actually has an internal buffer so performance is not a concern. `read_one` is essentially a [channel `recv`](https://github.com/chubei/dozer/blob/fix/consistency/dozer-log/src/reader.rs#L125). The `read_log_task` before this PR was implemented before the `LogReader` was modified to its current state.
2. If `CacheBuilderState::update` actually results in a new `building` cache, the `LogOperation`s from the last `LogReader` must not be passed to `CacheBuilderState::process`, so here once we get a `LogOperation` successfully, it's immediately passed to `CacheBuilderState::process`.

I'm not sure if I understand the "cache building workers" correctly. There's only one worker here that calls `CacheBuilderState::process`. It takes `&mut self` so we can't really use multiple workers.

Reply:
I was thinking something along the lines of (very much pseudocode):
```
tokio::spawn(async move {
    loop {
        let (rx, tx) = mpsc::channel(1);
        let cache = open_or_create_cache();
        tokio::spawn_blocking(|| build_log(rx, cache))
        while let Ok(msg) = read_one(){
             tx.send(msg)   
        }
        reconnect()
    }
})
```
So, control of which cache to build and which to serve is pulled outside of building the cache and the actual cache building happens in a purely sync task that doesn't know about there being multiple caches.

Reply:
I think I misunderstood you. You're talking about let the `cache` drop in `build_log` task, and create a new cache and task upon reconnectin?

Reply:
Above code can be written equivalently as following?

```
tokio::spawn_blocking(||
    loop {
        let cache = open_or_create_cache();
        while let Ok(msg) = tokio::block_on(read_one()) {
             build_one(cache, msg);
        }
        reconnect()
    }
)
```

This is indeed clearer. Let me do a quick refactor.